kayenta/kayenta-integration-tests/src/test/java/com/netflix/kayenta/configuration/EmbeddedPrometheusBootstrapConfiguration.java

```diff
@@ -13,66 +13,140 @@
 import org.testcontainers.utility.MountableFile;
 
 @Slf4j
-@Configuration
+@TestConfiguration
+@RequiredArgsConstructor
 public class EmbeddedPrometheusBootstrapConfiguration {
 
-  // Exposes host machine port to be used by prometheus container for scraping metrics from
-  // /prometehus endpoint
-  // See for more details:
-  // https://www.testcontainers.org/features/networking/#exposing-host-ports-to-the-container
-  static {
-    Testcontainers.exposeHostPorts(8081);
+  private static final int PROMETHEUS_INTERNAL_PORT = 9090;
+  private final Environment environment;
+
+  @Bean
+  public WaitStrategy prometheusWaitStrategy() {
+    return new HttpWaitStrategy()
+        .forPath("/status")
+        .forPort(PROMETHEUS_INTERNAL_PORT)
+        .forStatusCode(200);
   }
 
-  private static final int PORT = 9090;
+  private GenericContainer<?> prometheusContainer;
 
-  @Bean(name = "prometheusWaitStrategy")
-  public WaitStrategy prometheusWaitStrategy() {
-    return new HttpWaitStrategy().forPath("/status").forPort(PORT).forStatusCode(200);
+  @Bean
+  public ApplicationListener<ApplicationReadyEvent> prometheus(
+      ConfigurableEnvironment env, WaitStrategy prometheusWaitStrategy) {
+    return event -> {
+      int managementPort = waitForManagementPort();
+      exposeManagementPort(managementPort);
+
+      File prometheusConfigFile = createPrometheusConfigFile(managementPort);
+
+      prometheusContainer =
+          new GenericContainer<>("prom/prometheus:v2.10.0")
+              .withLogConsumer(containerLogsConsumer(log))
+              .withExposedPorts(PROMETHEUS_INTERNAL_PORT)
+              .withCopyFileToContainer(
+                  MountableFile.forHostPath(prometheusConfigFile.getAbsolutePath()),
+                  "/etc/prometheus/prometheus.yml")
+              .waitingFor(prometheusWaitStrategy)
+              .withStartupTimeout(Duration.ofSeconds(30));
+
+      startPrometheusServer(env);
+    };
   }
 
-  @Bean(name = "prometheus", destroyMethod = "stop")
-  public GenericContainer prometheus(
-      ConfigurableEnvironment environment, WaitStrategy prometheusWaitStrategy) {
-
-    GenericContainer container =
-        new GenericContainer("prom/prometheus:v2.10.0")
-            .withLogConsumer(containerLogsConsumer(log))
-            .withExposedPorts(PORT)
-            .withCopyFileToContainer(
-                MountableFile.forClasspathResource("/external/prometheus/prometheus.yml"),
-                "/etc/prometheus/prometheus.yml")
-            .waitingFor(prometheusWaitStrategy)
-            .withStartupTimeout(Duration.ofSeconds(30));
-    container.start();
-    Map<String, Object> env = registerEnvironment(environment, container.getMappedPort(PORT));
-    log.info("Started Prometheus server. Connection details: {}", env);
-    return container;
+  public void startPrometheusServer(ConfigurableEnvironment env) {
+    if (prometheusContainer != null && !prometheusContainer.isRunning()) {
```

Comment:
Do we need to check isRunning here?  The code here would be simpler if we can depend on prometheusContainer.start() to do the right thing if the container is already started.

Reply:
Will remove it

Reply:
removed here 

https://github.com/spinnaker/spinnaker/pull/7053/commits/9ab30db4a76dec4bc1afa3eb4710cca2c740f0f3